core/shared/src/main/scala/fs2/async/Promise.scala

```diff
@@ -0,0 +1,167 @@
+import cats.effect.{ Sync, Effect, IO }
+
+import fs2.Scheduler
+import fs2.internal.LinkedMap
+import java.util.concurrent.atomic.{AtomicBoolean,AtomicReference}
+
+import Promise._
+
+// TODO shall we move Effect and EC implicits back up here?
+// TODO scaladoc
+final class Promise[F[_], A] private [fs2] (ref: Ref[F, State[A]]) {
+
+  def get(implicit F: Effect[F], ec: ExecutionContext): F[A] =
+    F.delay(new MsgId).flatMap(getOrWait)
+
+  // TODO I prefer to not have `setAsync`, and leave forking at call site,
+  // since we can't exploit SyncRef lazy set directly
+  // I've replaced any setAsyncPure in fs2 with fork(setSync), which would
+  // probably reveal that there's a bunch of unnecessary forks that can be avoided
+  //
+  // NOTE: this differs in behaviour from the old Ref in that by the time readers are notified
+  // the new value is already guaranteed to be in place.
+  def setSync(a: A)(implicit F: Effect[F], ec: ExecutionContext): F[Unit] = {
+    def notifyReaders(r: State.Unset[A]): Unit =
+      r.waiting.values.foreach { cb =>
+        ec.execute { () => cb(a) }
+      }
+    // TODO if double set is allowed, this can be fast-path optimised
+    // to use SyncRef.set after the first `set`, however imho this pretty much
+    // never happens in practice, so it isn't really worth it
+    ref.modify2 {
+      case State.Set(_) => State.Set(a) -> F.unit //TODO allow double set, or fail? Allow. Useful for AsyncPull race, harmless for anything else (although perhaps the method itself could be simplified if set failed on second attempt)
+      case u @ State.Unset(_) => State.Set(a) -> F.delay(notifyReaders(u))
+    }.flatMap(_._2)
+  }
+
+  /** Like [[get]] but returns an `F[Unit]` that can be used cancel the subscription. */
+  def cancellableGet(implicit F: Effect[F], ec: ExecutionContext): F[(F[A], F[Unit])] = F.delay {
+    val id = new MsgId
+    val get = getOrWait(id)
+    val cancel = ref.modify {
+      case a @ State.Set(_) => a
+      case State.Unset(waiting) => State.Unset(waiting - id)
+    }.void
+
+    (get, cancel)
+  }
+
+  def timedGet(timeout: FiniteDuration, scheduler: Scheduler)(implicit F: Effect[F], ec: ExecutionContext): F[Option[A]] =
+    cancellableGet.flatMap { case (g, cancelGet) =>
+      scheduler.effect.delayCancellable(F.unit, timeout).flatMap { case (timer, cancelTimer) =>
+        fs2.async.race(g, timer).flatMap(_.fold(a => cancelTimer.as(Some(a)), _ => cancelGet.as(None)))
+      }
+    }
+
+  /**
+   * Runs `f1` and `f2` simultaneously, but only the winner gets to
+   * `set` to this `ref`. The loser continues running but its reference
+   * to this ref is severed, allowing this ref to be garbage collected
+   * if it is no longer referenced by anyone other than the loser.
+   *
+   * If the winner fails, the returned `F` fails as well, and this `ref`
+   * is not set.
+   */
+  def race(f1: F[A], f2: F[A])(implicit F: Effect[F], ec: ExecutionContext): F[Unit] = F.delay {
```

Comment:
Might want to make this `private[fs2]` or rename it to `unsafeRace` or something. Or change to `raceWith(f1: F[A], f2: F[A])(finished: Either[A,A] => F[Unit])(implicit ..): F[Unit]` to allow cooperative cancelation of the losing task.